sql: fix ColumnAccessExpr.Eval with NULL inner expression

mgartner opened the pull request:
#### sql: fix ColumnAccessExpr.Eval with NULL inner expression

Previously, `ColumnAccessExpr.Eval` would panic if the
`ColumnAccessExpr`'s inner expression evaluated to `NULL`, because it
attempted to cast this `NULL` to a `DTuple`. Now, if the inner
expression is `NULL`, `ColumnAccessExpr.Eval` returns `NULL`.

Fixes #78159

Release note (bug fix): A bug has been fixed that caused an internal
error when the inner expression of a column access expression evaluated
to `NULL`. For example, evaluation of the expression
`(CASE WHEN b THEN ((ROW(1) AS a)) ELSE NULL END).a` would error when
`b` is `false`. This bug has been present since version 19.1 or earlier.

#### sql: include tuple labels in types returned from typeCheckSameTypedTupleExprs

Previously, an expression that produced a tuple from several potential
values was typed as a tuple without any labels. This prevented a tuple's
column to be accessed by a label name.

For example, the expression below would result in the error "could not
identify column 'a' in record data type".

    SELECT (CASE WHEN true THEN (ROW(1) AS a) ELSE (ROW(2) AS a) END).a

Now, the labels of the first tuple are used in the type of the outer
expression.

Fixes #78515

Release note (bug fix): A bug has been fixed that caused an error when
accessing a named column of a labelled tuple. The bug only occurred when
an expression could produce one of several different tuples. For
example, `(CASE WHEN true THEN (ROW(1) AS a) ELSE (ROW(2) AS a) END).a`
would fail to evaluate. This bug was present since version 22.1.0.
Although present in previous versions, it was impossible to encounter
due to limitations that prevented using tuples in this way.

mgartner committed:
sql: fix ColumnAccessExpr.Eval with NULL inner expression

Previously, `ColumnAccessExpr.Eval` would panic if the
`ColumnAccessExpr`'s inner expression evaluated to `NULL`, because it
attempted to cast this `NULL` to a `DTuple`. Now, if the inner
expression is `NULL`, `ColumnAccessExpr.Eval` returns `NULL`.

Fixes #78159

Release note (bug fix): A bug has been fixed that caused an internal
error when the inner expression of a column access expression evaluated
to `NULL`. For example, evaluation of the expression
`(CASE WHEN b THEN ((ROW(1) AS a)) ELSE NULL END).a` would error when
`b` is `false`. This bug has been present since version 19.1 or earlier.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/78423)
<!-- Reviewable:end -->

msirek reviewed (approved):
Good fix!  Nice job reverse-engineering the test case.
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78423#-:-MywqKGg7E0V9ZeyZ7kC:brzq8t9)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cucaroach)



<!-- Sent from Reviewable.io -->

cucaroach reviewed (commented):
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78423#-:-MyxpBFP7L6jPVsxhjjQ:bmxilld)* status: :shipit: complete! 2 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

mgartner committed:
sql: include tuple labels in types returned from typeCheckSameTypedTupleExprs

Previously, an expression that produced a tuple from several potential
values was typed as a tuple without any labels. This prevented a tuple's
column to be accessed by a label name.

For example, the expression below would result in the error "could not
identify column 'a' in record data type".

    SELECT (CASE WHEN true THEN (ROW(1) AS a) ELSE (ROW(2) AS a) END).a

Now, the labels of the first tuple are used in the type of the outer
expression.

Fixes #78515

Release note (bug fix): A bug has been fixed that caused an error when
accessing a named column of a labelled tuple. The bug only occurred when
an expression could produce one of several different tuples. For
example, `(CASE WHEN true THEN (ROW(1) AS a) ELSE (ROW(2) AS a) END).a`
would fail to evaluate. This bug was present since version 22.1.0.
Although present in previous versions, it was impossible to encounter
due to limitations that prevented using tuples in this way.

mgartner commented:
The test case from the first commit revealed another bug, #75101. I've added another commit to address it, so please take another look @msirek @cucaroach.

mgartner commented:
Friendly ping 😃

msirek reviewed (approved):
The latest code revision looks good.  I tried out some test cases and all cases enforce that all THEN and ELSE branches conform to the type of the first THEN branch.  
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>  

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78423#-:-Mzzj06C5cuABceMRErf:bhb3cvx)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

mgartner commented:
TFTRs!

bors r+

cucaroach commented:
I guess once its merged reviewable goes into retirement?   Anyways :lgtm:
